src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxQueryBuilder.cs

```diff
@@ -87,14 +85,77 @@ internal class Parameters
         }
     }
 
-    private static IQueryMatch MaterializeWhenNeeded(IQueryMatch source)
+    public struct StreamingOptimization
     {
-        if (source is CoraxBooleanQueryBase cbq)
+        private readonly OrderMetadata _sortField;
+        public readonly bool OptimizationIsPossible;
+        public bool WhereClauseItemMatched;
+
+        public StreamingOptimization(OrderMetadata[] orderMetadata)
         {
-            source = cbq.Materialize();
+            //Conditions to consider streaming optimization:
+            //-sort only by one field
+            //-sort by Sequence/Number.
+            if (orderMetadata is null || orderMetadata.Length > 1 || orderMetadata is not
+                    [{FieldType: MatchCompareFieldType.Sequence or MatchCompareFieldType.Floating or MatchCompareFieldType.Integer}])
+            {
+                _sortField = default;
+                OptimizationIsPossible = false;
+                return;
+            }
+
+            _sortField = orderMetadata[0];
+            OptimizationIsPossible = true;
+        }
+        
+        public bool TrySetAsStreamingField(in CoraxBooleanItem cbi)
+        {
+            if (OptimizationIsPossible == false || cbi.Operation is not UnaryMatchOperation.Equals)
+                return false;
+         
+            if (WhereClauseItemMatched && OptimizationIsPossible)
+                ThrowFieldIsAlreadyChosen();
+
+            if (cbi.Field.Equals(_sortField.Field) == false)
+                return false;
+
+            // In case of TermMatch `order by` has to have exactly the same type as queried field.
+            // Since we store number as tuple of (String, Double, Long) we could write query like this:
+            // where Numeric = 1 order by Numeric as double. Where clause will match all documents where Numeric casted to long is equal 1 (e.g. 1.2, 1.3, 1.9) but order by will sort with precision.
+            bool matchOnType = cbi.Term switch
```

Comment:
Can we try to coerce it? 
So in the example above, we'll query on `1.0`, etc? Won't work the other way around, of course.
If I'm asking for `where Numeric = 1.5 order by Numeric`, we can't do this.

However, is there some indication that this is the case? We may need to start thinking about query plans hints, etc.